[Google Blockly] Convert XML to JSON and replicate domToBlockSpace behavior (experiment)

mikeharv opened the pull request:
# High-level context
This PR aims to add an experiment that will allow us to begin converting saved XML sources to JSON before loading blocks to a Google Blockly workspace. For more information about why we are interested in converting from XML to JSON, see the [JSON Serialization for saving user Blockly workspace state](https://docs.google.com/document/d/1V367MAAYU0JXETVX6l48_6HW2h93WDzdljJc1ohLZPc/edit#heading=h.3i8gd730jmak) dev doc.

## How we load blocks from XML
In all Blockly labs except Music Lab, we are doing using Blockly's XML system to save and load blocks.
Whether we are using XML or JSON, Blockly provides the functionality to both serialize a workspace (so that we can save it) and deserialize any saved state we give it (ie. load the blocks onto the workspace). In either case, our primary role is to handle the storage and retrieval of the data. Additionally, we also perform some custom actions during the loading process:
1. We override `domToBlock` so that we can create "unknown blocks" instead of crashing if any errors occur.
2. As we create blocks, we parse any X or Y coordinates set in the XML. We store the coordinates in an array alongside the blocks so that we can position them.
3. To position the blocks, we create a `cursor` object that we we use to position each block that doesn't have coordinates associated with it. This has to occur after loading blocks as the cursor is "repositioned" after each placement, based on the rendered size of each block, including SVG frames like we use for function definitions. 
4. In CDO Blockly labs, we reposition the blocks in two passes, saving the invisible blocks for last so that big gaps are not created between visible blocks. _While invisible blocks are not supported in Google Blockly, this logic was replicated in `cdoXml.js` in 2020 because we thought it might be supported._

For Google Blockly labs, all of this logic presently occurs in overrides of `Blockly.Xml.domToBlock` and a custom `Blockly.Xml.domToBlockSpace` function. The latter was created to replicate the version found in our CDO Blockly fork.

## Converting XML to JSON
With this new feature, we're interested in seeing if we can avoid loading blocks into our rendered main workspace from XML entirely. To accomplish this, we can use a temporary, headless (un-rendered) workspace to convert the blocks:
* if we don't already have JSON state, retrieve the XML
* create a temporary headless workspace
* use `domToBlockSpace` to load the XML blocks to the headless workspace *
* serialize the workspace to JSON
* dispose of the headless workspace

* Our `domToBlockSpace` has never been compatible with headless workspaces, so we need to change that. The latest commit moves any logic that is explicitly meant for rendered workspaces out of `domToBlockSpace` and into various new helpers in `cdoSerializationHelpers.js`.
From this point, we will be guaranteed to have JSON state to load which Blockly can handle. Note that we will likely need to support this flow _forever_ due to the vast number of projects/levels/tests that only have XML sources. 

## Replicating cdoXml behavior for JSON sources
Above, I listed four specific processes we use to manipulate blocks that we load from XML. I'll summarize them again here:
1. Create an unknown block if there is an error creating a single block
2. Track X/Y coordinates that are specified in the XML and move the rendered blocks to those coordinates.
3. Manually reposition blocks that don't have specified coordinates using a "cursor" object.
4. Sort the list of blocks based on visibility before positioning them.

We actually can skip the 4th item above as Google Blockly does not support invisible blocks. For the other items, the approach I used is detailed below.

### Unused Blocks
The JSON serialization system is made up of multiple serializers. There are built-in serializers for blocks and variables, and you can also register additional serializers. [[doc](https://developers.google.com/blockly/guides/configure/web/serialization)]

By registering a custom block serializer, we can continue to check for errors for each block and be ready to make "unknown blocks" in those instances. Because we do this during deserialization, we only need to extend the serializer class's `load` function.
* Existing XML implementation: https://github.com/code-dot-org/code-dot-org/blob/4c79d538ef25092e73f106d710e3a60f38e0a7e1/apps/src/blockly/addons/cdoXml.js#L5-L35
* New JSON block serializer: https://github.com/code-dot-org/code-dot-org/pull/52072/files#diff-a0b7c03f80568d7617253e4fe3aab1f6bf815bad50e80d8a9e3e6d7c0edc0d33
* Original JSON block serializer: https://github.com/google/blockly/blob/399bd650a69f50843f2b46a9907a8dce826f6b99/core/serialization/blocks.ts#L673-L738

### Getting x/y coordinates from XML
We will continue to call our custom `domToBlockSpace` to parse the xml. This function returns an array of "block" objects that include 1) the Blockly block, 2) the x position parsed from the xml, and 3) the y position parsed from the xml.
Before passing the newly converted JSON state back to Blockly to load, we will add write these x/y coordinate values into the JSON state. A map of block IDs referencing the serialized blocks is used to accomplish this.
* Existing XML implementation: https://github.com/code-dot-org/code-dot-org/blob/4c79d538ef25092e73f106d710e3a60f38e0a7e1/apps/src/blockly/addons/cdoXml.js#L47
* Implementation of new `addPositionsToState` function: https://github.com/code-dot-org/code-dot-org/pull/52072/files#diff-c3fddd86e1a595d93820a1c4c99988482427180fb77d33c2b7fd348e70c9e298R13-R45

### Manually positioning blocks 
If we don't have x/y coordinates for a block, Blockly would create it at `(0, 0)` (LTR) or `(viewWidth, 0)` (RTL). We want to move them to new locations determined by `cursor`. 
Prior to this PR, this happens as part of `domToBlockSpace`, even though this is a block manipulation and not really related to the XML any more. 
* XML implementation: https://github.com/code-dot-org/code-dot-org/blob/4c79d538ef25092e73f106d710e3a60f38e0a7e1/apps/src/blockly/addons/cdoXml.js#L67-L123
* New `positionBlocks` function: https://github.com/code-dot-org/code-dot-org/pull/52072/files#diff-c3fddd86e1a595d93820a1c4c99988482427180fb77d33c2b7fd348e70c9e298R47-R98


## Links

- [JSON Serialization for saving user Blockly workspace state](https://docs.google.com/document/d/1V367MAAYU0JXETVX6l48_6HW2h93WDzdljJc1ohLZPc/edit#heading=h.3i8gd730jmak) dev doc
- Jira - [[Google Blockly] Begin saving student code in JSON format in addition to XML](https://github.com/code-dot-org/code-dot-org/pull/52072/files)

## Testing story

I added a UI test that is a modified version of our existing `star_labs/blocklayout.feature`: https://github.com/code-dot-org/code-dot-org/pull/52072/files#diff-bca57f73f75f1c7b295e642c5b6c0cdff63330fece2623196656897029cc3eba
Both versions of this test are designed to test our ability to layout blocks correctly when give some arbitrary XML. The new version runs using the experiment flag in order to test the new pathway end-to-end. (It will only succeed if we can convert the XML to JSON and then position the blocks correctly.)

In addition, I tested a bunch of other scenarios, primarily using Flappy and Dance:
- Loading various existing projects, including with unknown blocks, unused blocks, function definitions, and event handlers placed at specific coordinates
- Loading hard-coded XML strings that include blocks with/without coordinates, blocks of undefined types, disconnected blocks

## Deployment strategy

All of the new features are placed behind an experiment flag just in case there are any concerns that this is too big of a shift to make. We have already tested that the XML and JSON formats are practically equivalent, but this reproduces a lot of positioning logic that was done in another way before. 
If we can try the experiment internally or a bit, we can consider lifting the experiment altogether and making this new flow for all users.

## Follow-up work

- Remove the experiment flag
- Begin saving JSON in parallel with the XML and begin loading it directly when it's present

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  6.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

fisher-alice commented on apps/src/blockly/addons/cdoBlockSerializer.js:
```diff
@@ -0,0 +1,38 @@
+import GoogleBlockly from 'blockly/core';
+
+const unknownBlockState = {type: 'unknown', enabled: false};
+const getRecordUndo = function () {
```
Could you explain what the purpose of this function is?

mikeharv replied:
It is telling `append` that we should log the action to the undo log, so that users can undo things like pasting blocks. However, I just realized we can use `Blockly.Events.getRecordUndo` directly and not need to redefine it ourselves.

fisher-alice replied:
Oh - interesting! Thanks for explaining.

fisher-alice commented on apps/src/blockly/addons/cdoSerializationUtils.js:
```diff
@@ -0,0 +1,109 @@
+function addPositionsToState(xmlBlocks, blockIdMap) {
+  xmlBlocks.forEach(xmlBlock => {
+    const block = blockIdMap.get(xmlBlock.blockly_block.id);
+    if (block && xmlBlock.x && xmlBlock.y) {
```
Is there ever a case when a block has only either `x` or `y` value but not both?

mikeharv replied:
Not that I can find! But I'm happy to adjust so that each is set separately.

mikeharv replied:
Updated logic:

```
    if (block) {
      // Do not change block values if xmlBlock values are NaN (unspecified in XML)
      block.x = xmlBlock.x || block.x;
      block.y = xmlBlock.y || block.y;
    }
```
Now, if a block only has one coordinate set in the XML, it will get used. This does mean the other coordinate will be 0.  Below, the block XML for `when hit the ground` includes `x = "250"` while `when hit an obstacle` includes `x="250" y="250"`. All are blocks have neither and re-positioned using the cursor.

![image](https://github.com/code-dot-org/code-dot-org/assets/43474485/a0592c30-c695-4087-81a3-92c5866c07b7)

fisher-alice replied:
In the case above for `when hit the ground`, would it be possible to add a cushion for the y value? Or maybe this is a way to remind levelbuilders to add both x and y coordinates? 
Can you remind me for which Google Blockly labs is there a reorganization of the blocks and levelbuilders are unable to set the blocks to a certain location? I remember Emma asking about this.

mikeharv replied:
Emma was asking about Sprite Lab, where it looks like we don't currently save block X/Y coordinates. This will start to happen by default once it's on Google Blockly, fortunately.

I'll look into the cushion idea. It will complicate the logic a bit. If both coordinates are 0, we let the cursor position it. So we'd only want to add the cushion to the unspecified value if the other was specified.

fisher-alice commented on apps/src/blockly/addons/cdoSerializationUtils.js:
```diff
@@ -0,0 +1,109 @@
+import BlockSvgUnused from './blockSvgUnused';
```
Just curious why these functions not included in `cdoUtils`? Also, I expected`cdoSerializationUtils` to contain functions that serialized the workspace blocks. Since these functions are converting from one way to serialize into another way to serialize, or repositioning blocks, maybe name them something like `cdoSerializationHelpers`? 
Also, this is out of scope of this PR but our `addons` directory seems to be growing fairly quickly - I wonder if we should think about adding subfolders for organization?

mikeharv replied:
That growth is exactly why I pulled these new functions out of cdoUtils! I wasn't sure of the best name, but agree with your interpretation. I'll rename to `cdoSerializationHelpers`. I agree that the `addons` folder itself is getting large. I'd love to see if we could sync as an internal Blockly team to figure out the best way forward.

fisher-alice replied:
Yes - sounds good on reorganizing the `addons` folder. I'll add a jira ticket for this.

fisher-alice replied:
https://codedotorg.atlassian.net/browse/SL-881 - the ticket also includes a link to our Slack discussion when we discussed our wrappers and 'shims'.

fisher-alice commented on dashboard/test/ui/features/star_labs/blocklayoutjson.feature:
```diff
@@ -0,0 +1,16 @@
+Feature: Block auto-layout after JSON conversion
+
+Background:
+  Given I am on "http://studio.code.org/flappy/10?noautoplay=true&enableExperiments=blocklyJson"
```
Thanks for including a UI test!

fisher-alice commented on apps/src/blockly/addons/cdoSerializationUtils.js:
```diff
@@ -0,0 +1,109 @@
+}
+
+/**
+ * Converts XML serialization to JSON using a temporary unrendered workspace.
```
I think this comment needs to be updated.

mikeharv replied:
Thanks for catching! Will correct.

fisher-alice reviewed (commented):
Great job, Mike! Just posed a couple questions to make sure I am understanding the work of this PR - thanks!

bencodeorg commented on apps/src/blockly/addons/cdoBlockSerializer.js:
```diff
@@ -0,0 +1,35 @@
+
+    for (const blockState of blockStates) {
+      try {
+        GoogleBlockly.serialization.blocks.append(blockState, workspace, {
```
Nit: does this work without the `GoogleBlockly.serialization.blocks.` prefix?

bencodeorg commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,110 @@
+ * @param {Blockly.Workspace} workspace - the current Blockly workspace
+ * @param {Map<String, Object>} blockIdMap - a map of ids (keys) and blocks (values)
+ */
+export function positionBlocks(workspace) {
```
Most (all?) of this logic is from cdoXml, yeah? Could you add a comment to that effect?

bencodeorg commented on apps/src/blockly/addons/cdoBlockSerializer.js:
```diff
@@ -0,0 +1,35 @@
+      } catch (e) {
+        console.warn(`Creating "unknown block". ${e.message}`);
+        GoogleBlockly.serialization.blocks.append(
+          Object.assign(unknownBlockState, {x: blockState.x, y: blockState.y}),
```
I'm not sure if it matters, but I think what you're doing here is modifying the `unknownBlockState` object (which I think should be a template/unchanged), rather than copying its properties into a new object. I think probably more appropriately written as:

`{...unknownBlockState, x: blockState.x, y: blockState.y}`

mikeharv replied:
Oh good call! I think that this conveniently worked, but agree that there's no need to add the x/y values to the template. Updating now.

thomasoniii commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,110 @@
+  const blockIdMap = new Map();
+  stateToLoad.blocks.blocks.forEach(block => {
+    blockIdMap.set(block.id, block);
+  });
```
as a non-blocking suggestion, this is a perfect place to use a reduce:

`const blockIdMap =
  stateToLoad.blocks.blocks.reduce( (map, block) => map.set(block.id, block), new Map() )`

`Map.set()` conveniently returns the `Map` object itself, so we can use it as the return in the callback.

fisher-alice replied:
Cool!

bencodeorg commented on apps/src/blockly/addons/cdoUtils.js:
```diff
@@ -2,9 +2,26 @@ import {ToolboxType, CLAMPED_NUMBER_REGEX, DEFAULT_SOUND} from '../constants';
+ * @param {xml} xml - workspace serialization, current/legacy format
+ * @param {*} stateToLoad - modern workspace serialization, may not be present
+ */
 export function loadBlocksToWorkspace(workspace, xml, stateToLoad) {
```
Is this list of arguments copied from google blockly, or is it our own? I'm a little confused at the names "xml" and "stateToLoad"...is `stateToLoad` the same thing as "xml" just in a json format?

mikeharv replied:
This function is our own as are the parameters.  Throughout Google Blockly core, they use `state` to refer to the JSON-formatted serialization that is being loaded. But yes, both `xml` and `stateToLoad` are different formats of the same thing. Open to suggestions for naming here!

bencodeorg replied:
Ok cool if it's consistent with them, good with me. Do they use `xml` to refer to what we call `xml` here? 

We could align the naming in this function (`xmlState`, `jsonState`) if we want to make clear to future engineers that it's the same thing...whatever you think is better.

mikeharv replied:
I like that suggestion (xmlState, jsonState).

And yes, like us, they use `xml` for the older XML functions.

bencodeorg commented on apps/src/blockly/addons/cdoXml.js:
```diff
@@ -59,6 +59,12 @@ export default function initializeBlocklyXml(blocklyWrapper) {
       });
     });
 
+    // Below this point, we will be repositioning blocks on a rendered workspace.
```
Would it be possible to add `reposition` as an optional third argument to this function, rather than relying on this early return? Seems like something we could accidentally break.

mikeharv replied:
Looking at this again... I think that the code below this point doesn't really need to be contained in `domToBlockSpace`, since at this point, the blocks are on the workspace. Instead, what if we moved all the repositioning logic to a separate function, and then updated our calls like this:

_Before:_
`Blockly.Xml.domToBlockSpace(workspace, xml);`

_After:_
```
    const xmlBlocks = Blockly.Xml.domToBlockSpace(workspace, xml);
    if (workspace.rendered) {
      positionBlocks(xmlBlocks, workspace);
    }
```

bencodeorg replied:
Even better!

bencodeorg commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,110 @@
+export function convertXmlToJson(xml) {
+  const tempWorkspace = new Blockly.Workspace();
+
+  // domToBlockSpace returns an array of "block" objects with the following property:
```
typo: properties

bencodeorg reviewed (commented):
A few small comments, but I think the general direction looks great! And the UI test you've added seems like a good way we could start to add some more test coverage here 👍

thomasoniii commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,110 @@
+  }
+  const metrics = workspace.getMetrics();
+  const contentWidth = metrics.contentWidth;
+  const viewWidth = metrics ? metrics.viewWidth : 0;
```
This logic doesn't feel right, could you elaborate?
Line 59 assumes that the `metrics` object exists and snags `contentWidth` from it.
But line 60 then checks to see if the `metrics` object exists. It _has_ to exist at this point (if it didn't then line 59 would've blown up), so I think this ternary will never be false and it'll always stick in `metrics.viewWidth`, which may be undefined/null/some other value, I suppose.

Should it be something more like
`const metrics = workspace.getMetrics() || {}`
and then line 60 be `const viewWidth = metrics.viewWidth ?? 0` ?

What's `getMetrics` return type? If it's guaranteed to have an object, we may need the default at all. Alternatively, it may be good to populate a default object with extra values, i.e. `{ viewWidth : 0, contentWidth : 0}`

mikeharv replied:
Good call! `getMetrics()` should always return an object. The `viewWidth` line was copied from elsewhere, and I think we just were checking for the existence of `metrics` out of an abundance of caution. I didn't think to follow this same pattern when I added the `contentWidth` line. To be on the safe side, I went with your suggestion of a populated default object with some 0 values.

thomasoniii commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,110 @@
+    x: padding,
+    y: padding,
+  };
+  if (workspace.RTL && viewWidth) {
+    // Position the cursor from the right of the workspace.
+    cursor.x = viewWidth - padding;
+  } else if (workspace.RTL && !viewWidth) {
+    // Position the cursor from the right of the block.
+    cursor.x = contentWidth - padding;
+  }
```
Again, non-blocking suggestion, but this could be crunched down into:
`
if (workspace.RTL) {
  cursor.x = (viewWidth || contentWidth) - padding
}
`
I'm not super stoked about it since I'm not crazy about the embedded `||` in the subtraction. Maybe put that in a separate variable?
`
if (workspace.RTL) {
  const width = viewWidth || contentWidth
  cursor.x = width - padding
}
`

Again, there's nothing wrong with the way it currently is, and I can see arguments against the more compact version. I'm just trying to get rid of the slightly redundant logic, though this is a case of diminishing returns on it, I feel.

If `viewWidth` could use `contentWidth` as its default it'd clean it up even more, but I don't think that's the case.

Also - is there any way to type multi-line code blocks in a github comment? :-(

bencodeorg replied:
@thomasoniii you can do multi-line code blocks with triple ticks (like ```) on their own lines preceding/following the lines containing your code:

```
if (foo) {
  bar()
}
```

mikeharv replied:
Could also do a ternary like:
```
const width = viewWidth || contentWidth;
cursor.x = workspace.RTL ? width - padding : cursor.x;

```
Is that more readable @thomasoniii ?

mikeharv replied:
Ooh, how's this?

```
  const metrics = workspace.getMetrics() || {contentWidth: 0, viewWidth: 0};
  const contentWidth = metrics.contentWidth;
  const viewWidth = metrics.viewWidth;
  const padding = viewWidth ? 16 : 0;
  const width = viewWidth || contentWidth;

  // The "cursor" object tracks a position on the workspace that starts in the top
  // corner, then moves below each block that is manually repositioned.
  let cursor = {
    x: padding,
    y: padding,
  };
  // If the workspace is RTL, horizontally mirror the starting position.
  cursor.x = workspace.RTL ? width - cursor.x : cursor.x;
```

fisher-alice replied:
I like this latest version - simple and clear.

thomasoniii replied:
I like it! With one further suggestion to use destructuring to avoid the interim variables:

```
const { contentWidth = 0, viewWidth = 0} = workspace.getMetrics();
```

The logic here is sneaky - if `getMetrics` is always returning an object, then we'll never fall past the `||` clause to get to the populated defaults. So instead, give them initial values in case they're not assigned.

You could make it `workspace.getMetrics() || {}` and keep the defaults in the destructuring, but if `getMetrics` is always returning an object anyway, I'd say that it just adds clutter and isn't strictly necessary.

mikeharv replied:
Very cool, thanks for the tips!

mikeharv replied:
Very cool, thanks for the tips!

mikeharv commented:
In a recent commit (https://github.com/code-dot-org/code-dot-org/pull/52072/commits/86bcf61aef129452c3ecfd35dba008f5ec9b2916) I performed a refactor of some existing logic from cdoXml and some new similar logic in cdoSerializationHelpers.

Previously, `domToBlockSpace` included some positioning logic that wasn't strictly related to XML parsing and was isntead about moving Blockly blocks that had already been rendered. This code didn't make sense to include here as `domToBlockSpace` can also be used with headless (non-rendered) workspaces. Rather than rely on an early return, I moved this out to `positionBlocksOnWorkspace` (shared logic for getting ready to reposition all blocks), and `positionBlockLegacy` (logic that was unique to the path found in cdoXml. There is also a `positionBlock` that covers the new simpler path that we can use to move a block if it was created from a JSON serialization. 

While performing this refactor, I also removed a couple things out right:

- [`reorderBlocks`](https://github.com/code-dot-org/code-dot-org/pull/52072/commits/86bcf61aef129452c3ecfd35dba008f5ec9b2916#diff-1deb52d0698ae48e74781e441207cbeb1ffffd69fc70293a3dce035cb216e6aeL150-L163) -  This was [added by me](https://github.com/code-dot-org/code-dot-org/pull/48631) in November of 2022 to handle an uncertainty about what would happen if Dance Party functions that were defined via modal editor were suddenly loaded into a level that no longer supported the modal editor. I've since learned that CDO Blockly was already sorting these after other visible blocks, so this shouldn't be necessary. 
- [sorting based on Block visibility](https://github.com/code-dot-org/code-dot-org/pull/52072/commits/86bcf61aef129452c3ecfd35dba008f5ec9b2916#diff-1deb52d0698ae48e74781e441207cbeb1ffffd69fc70293a3dce035cb216e6aeL127-L137). This was added in 2020 when we thought mainline Blockly might be able to support invisible blocks. It doesn't, and `isVisible()` has always just returned true: https://github.com/code-dot-org/code-dot-org/blob/mike/replicate-domToBlockSpace-for-json/apps/src/blockly/googleBlocklyWrapper.js#L386-L390
- [`blockToDomWithXY`](https://github.com/code-dot-org/code-dot-org/pull/52072/commits/86bcf61aef129452c3ecfd35dba008f5ec9b2916#diff-1deb52d0698ae48e74781e441207cbeb1ffffd69fc70293a3dce035cb216e6aeL145-L147) - We never call this. We also _do_ save block positions, contrary to the original comment.

fisher-alice commented:
> In a recent commit ([86bcf61](https://github.com/code-dot-org/code-dot-org/commit/86bcf61aef129452c3ecfd35dba008f5ec9b2916)) I performed a refactor of some existing logic from cdoXml and some new similar logic in cdoSerializationHelpers.
> 
> Previously, `domToBlockSpace` included some positioning logic that wasn't strictly related to XML parsing and was isntead about moving Blockly blocks that had already been rendered. This code didn't make sense to include here as `domToBlockSpace` can also be used with headless (non-rendered) workspaces. Rather than rely on an early return, I moved this out to `positionBlocksOnWorkspace` (shared logic for getting ready to reposition all blocks), and `positionBlockLegacy` (logic that was unique to the path found in cdoXml. There is also a `positionBlock` that covers the new simpler path that we can use to move a block if it was created from a JSON serialization.
> 
> While performing this refactor, I also removed a couple things out right:
> 
> * [`reorderBlocks`](https://github.com/code-dot-org/code-dot-org/pull/52072/commits/86bcf61aef129452c3ecfd35dba008f5ec9b2916#diff-1deb52d0698ae48e74781e441207cbeb1ffffd69fc70293a3dce035cb216e6aeL150-L163) -  This was [added by me](https://github.com/code-dot-org/code-dot-org/pull/48631) in November of 2022 to handle an uncertainty about what would happen if Dance Party functions that were defined via modal editor were suddenly loaded into a level that no longer supported the modal editor. I've since learned that CDO Blockly was already sorting these after other visible blocks, so this shouldn't be necessary.
> * [sorting based on Block visibility](https://github.com/code-dot-org/code-dot-org/pull/52072/commits/86bcf61aef129452c3ecfd35dba008f5ec9b2916#diff-1deb52d0698ae48e74781e441207cbeb1ffffd69fc70293a3dce035cb216e6aeL127-L137). This was added in 2020 when we thought mainline Blockly might be able to support invisible blocks. It doesn't, and `isVisible()` has always just returned true: https://github.com/code-dot-org/code-dot-org/blob/mike/replicate-domToBlockSpace-for-json/apps/src/blockly/googleBlocklyWrapper.js#L386-L390
> * [`blockToDomWithXY`](https://github.com/code-dot-org/code-dot-org/pull/52072/commits/86bcf61aef129452c3ecfd35dba008f5ec9b2916#diff-1deb52d0698ae48e74781e441207cbeb1ffffd69fc70293a3dce035cb216e6aeL145-L147) - We never call this. We also _do_ save block positions, contrary to the original comment.

This is really nice refactoring! So cool that we now pass the function `positionBlocks` based on which serialization was used. I know mainline Blockly uses the term 'legacy' to refer to xml pathway. But do we have a shared understanding that 'legacy' indicates the xml serialization? It seems like we're actively using xml and json at this point I wonder if `positionBlocksFromJson` and `positionBlocksFromXml` may make it easier to read at this point?

fisher-alice commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,169 @@
+ * @param {number} cursor.y - a y-coordinate for moving a block
+ */
+export function positionBlock(block, workspace, cursor) {
+  const verticalSpaceBetweenBlocks = 10;
```
Now that we have two versions of `positionBlocks`, would it make sense to create constants at the top of the file for these duplicates? Same with the `frameSvgTop`, etc below.

mikeharv replied:
Yes, definitely. I should be able to move any constants that aren’t based on workspace metrics.

bencodeorg commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,169 @@
+ */
+function addPositionsToState(xmlBlocks, blockIdMap) {
+  xmlBlocks.forEach(xmlBlock => {
+    const block = blockIdMap.get(xmlBlock.blockly_block.id);
```
Nit: could we distinguish these names a bit? `jsonBlock` rather than `block` here maybe?

bencodeorg commented:
Note to update PR description now that you're not doing the early return thing in headless workspaces (and to make sure the description is still accurate otherwise)

mikeharv commented:
One more new commit: [b7aab67](https://github.com/code-dot-org/code-dot-org/pull/52072/commits/b7aab67a0b7e739a1e34c4ca23a6a0bcb7d4c909)

* incorporated tips from @thomasoniii about better destructuring use to avoid interim variables
* incorporated feedback from @fisher-alice to better use constants. in fact, we already had some constants elsewhere for describing svg frame dimensions, so i've reused those.
* refactored the two variants of `positionBlock` so that the XML version is simply a helper that gets things in the right state to use the other one
* added more helpers with JSDocs to hopefully breakdown all of this behavior a little better and make it more maintainable!

bencodeorg commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,184 @@
+ * @param {number} cursor.y - a y-coordinate for moving a block
+ * @return {object} the cursor with updated coordinates
+ */
+export function positionBlockWtihCursor(block, cursor) {
```
Typo

fisher-alice commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,184 @@
+ */
+export function positionBlocksOnWorkspace(
+  workspace,
+  positionBlock = positionBlockWtihCursor,
```
Typo: `positionBLocksWithCursor` and line 95

mikeharv replied:
Thank you for catching! Fixed

bencodeorg commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,184 @@
+ * @param {number} cursor.y - a y-coordinate for moving a block
+ * @return {object} the cursor with updated coordinates
+ */
+export function positionBlockWithCursor(block, cursor) {
```
Does this need to be exported?

bencodeorg commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,184 @@
+    VERTICAL_SPACE_BETWEEN_BLOCKS + (blockHasFrameSvg ? SVG_FRAME_HEIGHT : 0);
+  return blockHeight + blockVerticalPadding;
+}
+/**
```
Nit: newline here

fisher-alice commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,184 @@
+ * @param {function} [positionBlock] - moves a single block using the current cursor coordinates
+ * @param {Array<block>} [blocks] - an array of block objects
+ */
+export function positionBlocksOnWorkspace(
```
I wonder if a comment may be helpful here for readers to understand that if `positionBlock` and `blocks` are passed, they are using the xml pathway as opposed to the json pathway which uses the default parameter values.

fisher-alice commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,184 @@
+
+const {HEADER_HEIGHT, MARGIN_BOTTOM, MARGIN_SIDE, MARGIN_TOP} = frameSizes;
+const SVG_FRAME_HEIGHT = HEADER_HEIGHT + MARGIN_TOP + MARGIN_BOTTOM;
+const SVG_FRAME_SIDE_PADDING = MARGIN_SIDE;
```
Nit: maybe use `SvgFrame` object with values for `HEIGHT`, `SIDE_PADDING`, and `TOP_PADDING`.

bencodeorg reviewed (approved):
The new experimental behavior looks good to me, and the general move to break apart block positioning and deserialization makes sense to me.

Given all the refactoring, I'd recommend one more pass through your manual testing list before merging. Know I've definitely made unexpected changes doing a "quick refactor" during a review process :)

fisher-alice commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,184 @@
+ * @param {Blockly.Block} block - the block being considered
+ * @return {boolean} - true if the block is at the top corner of the workspace
+ */
+function blockLocationUnset(block) {
```
The name of this function sounds like it may unset or change the block location instead of checking whether the block needs to be re-positioned - maybe considering something like `isBlockLocationUnset`?

fisher-alice commented on apps/src/blockly/addons/cdoSerializationHelpers.js:
```diff
@@ -0,0 +1,184 @@
+ * @param {number} [block.y] - a y-coordinate frmo the XML serialization
+ * @param {object} cursor - a location for moving a block
+ */
+export function positionBlockXmlHelper(block, cursor) {
```
Nice refactor!

fisher-alice reviewed (approved):
Thanks for iterating on this, @mikeharv. I agree with @bencodeorg that the scope of this PR ended up larger than initially planned with the refactoring, removal of unnecessary code, etc.
As someone who was 'on the ride' from the beginning,  I understand how the scope grew. But can definitely see how folks looking back on this PR may have a challenging time following all the updates.

Overall - awesome work Mike! I learned a ton from you, @bencodeorg  and @thomasoniii while reviewing.
